charon/src/transform/index_to_function_calls.rs

```diff
@@ -36,9 +36,26 @@ impl<'a> Transform<'a> {
         let mut var_id = p.var_id;
         let mut proj = Vec::new();
         for pe in p.projection.clone().into_iter() {
-            if pe.is_index() {
-                let (index_var_id, buf_ty) = pe.to_index();
-
+            if let ProjectionElem::Index(operand, buf_ty) = pe {
+                let index_var_id = match &operand {
+                    // If we have a place, just use directly its identifier
+                    Operand::Copy(place) => place.var_id,
+                    // Otherwise, we push a new local and an `Assign` statement
+                    Operand::Const(ConstantExpr { ty, .. }) => {
+                        let var_id = self.fresh_var(None, ty.clone());
+                        let place = Place {
+                            var_id,
+                            projection: vec![],
+                        };
+                        let content = RawStatement::Assign(place, Rvalue::Use(operand));
+                        let span = self.span.unwrap();
+                        self.statements.push(Statement { span, content });
+                        var_id
+                    }
+                    Operand::Move(_) => {
+                        panic!("Unexpected `Oparand::Move` on a `ProjectionElem::Index`")
+                    }
+                };
```

Comment:
`index_var_id` is only used below as `let arg_index = Operand::Copy(Place::new(index_var_id))`. You don't need to add an `Assign` at all, you can simply do `let arg_index = operand;`.

Reply:
Oh, right, fantastic! Thanks!